### ISSUE TITLE
Torch SD-based models tensor invalid for input size

### ISSUE THREAD

**Issue by pattontim** [huggingface/safetensors#95]
There might be a slight discrepancy between the loading and saving process in safetensors. When loading a SD-based model like sd-1.4 packaged into a PyTorch checkpoint, we'll call it sd-v1-4.ckpt. We can package its state_dict and discard the torch format.

Packaging as safe_tensors

```
    sf_filename = "sd-v1-4.safetensors"
    filename = "sd-v1-4.ckpt"

    loaded = torch.load(filename)
    loaded = loaded['state_dict']

    # appears to pop nothing in this case
    shared = shared_pointers(loaded)
    for shared_weights in shared:
        for name in shared_weights[1:]:
            loaded.pop(name)

    loaded = {k: v.contiguous() for k, v in loaded.items()}

    save_file(loaded, metadata={"format": "pt"})

    check_file_size(local, filename)
```

Loading the tensors

`    load_file('sd-v1-4.safetensors', device='cpu')`

Results in error: 

 File "venv\lib\site-packages\safetensors\torch.py", line 99, in load_file
    result[k] = f.get_tensor(k)
RuntimeError: shape '[1280, 1280, 3, 3]' is invalid for input of size 7290352

Expected behaviour: safetensors fails while trying to save unexpected tensor data or creates tensors which can be loaded
Affected version: safetensors=2.4.0, torch=1.12.1+cu113

ckpt size: 
3.97 GB 4,265,381,888 bytes (4,265,380,512 bytes)
safetensor size:
3.97 GB 4,265,148,416 bytes (4,265,146,304 bytes)
SHA fe4efff1e174c627256e44ec2991ba279b3816e364b49f9be2abc0b3ff3f8556

Using pruned version of [CompVis/stable-diffusion-v-1-4-original](https://huggingface.co/CompVis/stable-diffusion-v-1-4-original)

Apologies if this is already fixed with addition of more dtypes. Will try to get more info by running through check output and debug info of this specific tensor

**Comment by Narsil** [comment]
Hi @pattontim ,

Thanks for the report.

So far I have reproduced and the files seem to match 1:1.

Here is a script I propose:

```python
from huggingface_hub import hf_hub_download
from safetensors.torch import load_file
import torch

pt_filename = hf_hub_download(repo_id="CompVis/stable-diffusion-v-1-4-original", filename="sd-v1-4.ckpt")
pt_loaded = torch.load(pt_filename)["state_dict"]
sf_filename = hf_hub_download(repo_id="CompVis/stable-diffusion-v-1-4-original", filename="model.safetensors", revision="refs/pr/224")
sf_loaded = load_file(sf_filename)

for k, v in pt_loaded.items():
    pt_tensor = v
    sf_tensor = sf_loaded[k]

    if not torch.allclose(pt_tensor, sf_tensor):
        raise Exception("Difference")

print("Seems everything is ok.")
```

**Comment by Narsil** [comment]
The error doesn't have a stacktrace but it's probably linked to some other part of the pytorch checkpoint being missing/improperly created.

What is your hardware ?

I tried to reproduce but couldn't with https://github.com/huggingface/safetensors/blob/main/bindings/python/convert.py even with your pytorch version.

**Comment by Narsil** [comment]
Also too the liberty to open a PR against stable-diffusion-webui since I saw you opened an issue there : https://github.com/AUTOMATIC1111/stable-diffusion-webui/pull/4930 (I also like it ! :D )

**Comment by pattontim** [comment]
My hardware is a GTX3060 and I'm using it with the SAFETENSORS_FAST_GPU

**Comment by pattontim** [comment]
My hardware is a GTX3060 and I used SAFETENSORS_FAST_GPU when creating the file and when loading the file. I tried it again without this flag at creation time and without the flag at load time and it still didn't work.

**Comment by pattontim** [comment]
It appears to successfully load the file if I set device='cuda'. My CPU is amd fx6350 and only half of my RAM is used at peak

**Comment by Narsil** [comment]
> I'm loading the model when swapping from one model with xformers applied, to the safetensors file.

I don't know what xformers does, but it's low level enough to screw things up.
Do you mind sharing a simple reproducing script ?
Did you try the script I suggested ? Does it work ?

**Comment by pattontim** [comment]
> > I'm loading the model when swapping from one model with xformers applied, to the safetensors file.
> 
> I don't know what xformers does, but it's low level enough to screw things up. Do you mind sharing a simple reproducing script ? Did you try the script I suggested ? Does it work ?

I should have reworded, xformers is applied to the model loaded before the safetensor. However I eliminated this being relevant by restarting my PC and just using your script but with the safetensor I created instead. I used your script but instead load the safetensor I created with the script in the OP

**Comment by pattontim** [comment]
> 
> **Could you share:**
> 
>     * Your windows version (And if WSL or not)
> 
>     * Your python version
> 
>     * You environment (conda ,pyenv, virtualenv, or none)
> 

Windows 11 Pro 22621.674

Python 3.10.6 (tags/v3.10.6:9c7b4bd, Aug  1 2022, 21:53:49) [MSC v.1932 64 bit (AMD64)]

venv

**Comment by pattontim** [comment]
gpt2 loads fine. After adding I64 to the dtypes, when running with the sd safetensor I originally created, I get the error

`venv "D:\test\venv\Scripts\Python.exe"
Traceback (most recent call last):
  File "D:\test\convert-pure.py", line 38, in <module>
    main()
  File "D:\test\convert-pure.py", line 33, in main
    weights = load_file(filename, device)
  File "D:\test\convert-pure.py", line 17, in load_file
    return {name: create_tensor(storage, info, offset) for name, info in metadata.items() if name != "__metadata__"}
  File "D:\test\convert-pure.py", line 17, in <dictcomp>
    return {name: create_tensor(storage, info, offset) for name, info in metadata.items() if name != "__metadata__"}
  File "D:\test\convert-pure.py", line 28, in create_tensor
    return torch.asarray(storage[start + offset : stop + offset], dtype=torch.uint8).view(dtype=dtype).reshape(shape)
RuntimeError: shape '[1000]' is invalid for input of size 500
Press any key to continue . . .`

half precision?

**Comment by pattontim** [comment]
Deleted a post where I accidentally put I64 for I32.

Loads both the gpt and previously-erring safetensors file fine with the pure python approach. I added to DTYPES `"I64":torch.int64, "I32":torch.int32`

**Comment by Narsil** [comment]
So your issue is fixed ?

> Loads both the gpt and previously-erring .safetensors file fine and outputs the keys with the pure python approach. I added to DTYPES "I64":torch.int64, "I32":torch.int32

Both should already exist : https://github.com/huggingface/safetensors/blob/main/bindings/python/py_src/safetensors/torch.py#L149

**Comment by pattontim** [comment]
> So your issue is fixed ?
> 
> > Loads both the gpt and previously-erring .safetensors file fine and outputs the keys with the pure python approach. I added to DTYPES "I64":torch.int64, "I32":torch.int32
> 
> Both should already exist : https://github.com/huggingface/safetensors/blob/main/bindings/python/py_src/safetensors/torch.py#L149

As in I ran your gist https://gist.github.com/Narsil/3edeec2669a5e94e4707aa0f901d2282 and it worked. Maybe it will work in dev if the torch fixes since release 0.2.4?

**Comment by Narsil** [comment]
> As in I ran your gist https://gist.github.com/Narsil/3edeec2669a5e94e4707aa0f901d2282 and it worked. 

Very good signal, the issue is in the loading, not in the serializing.

> Maybe it will work in dev if the torch fixes since release 0.2.4?

I don't think much has changed since. Though could you try to install from source to check ?

You need rust https://rustup.rs/

Then 

```bash
cd bindings/python
pip install setuptools_rust
pip install -e .
```

**Comment by pattontim** [comment]
I want to say its not fixed on the dev build but I can't say for certain. When that runs it works as expected until the first torch model is loaded, then I get a windows error that python has stopped working. The program continues to load second model over time and then the program terminates without a message. No "Seems everything is ok.". Just a blank terminal

**Comment by Narsil** [comment]
Can you upload the serialized file somewhere ?

**Comment by pattontim** [comment]
> Can you upload the serialized file somewhere ?

Is it necessary? Is the SHA256 of model.safetensors on your sd1.4 PR different than e57901186bb65c5b7b9fce118dd221bd646fdcc0a8ab34dfdc25ead5bd11fb59 ?

**Comment by Narsil** [comment]
https://huggingface.co/CompVis/stable-diffusion-v-1-4-original/discussions/224/files

No it's not.

So this 

```python
from huggingface_hub import hf_hub_download
from safetensors.torch import load_file

sf_filename = hf_hub_download(repo_id="CompVis/stable-diffusion-v-1-4-original", filename="model.safetensors", revision="refs/pr/224")
sf_loaded = load_file(sf_filename)
```

Fails on your machine too ?
If yes, it doesn't fail with https://gist.github.com/Narsil/3edeec2669a5e94e4707aa0f901d2282 (with `gpt2` replaced by the proper `repo_id` and `revision`), correct ?

That would be extremely weird, since both codes executes pretty much the same thing.

> that python has stopped working

Multiple potential causes, being out of memory is classic though

> The GPU continues to load second model over time 

All the code I shared is supposed to load things on CPU, if things are loaded on GPU something is wrong. (The goal here is to identify the bug so we need to remove as much code as possible, just putting things on CPU should be enough to gauge if the loaded tensors are correct.)

**Comment by Narsil** [comment]
Okay, thanks to @mfuntowicz which has a windows machine we were able to figure out.

Turns out it IS because of windows that there is an issue.
https://doc.rust-lang.org/stable/std/ffi/type.c_long.html  `c_long` is a `i32` on Windows platform meaning the tensor slice we're taking is overflowing and taking a wrong slice.

This IMO is a PyO3 issue, which I will report, and will provide a hotfix soon.

Thank you so much for helping on this.
